src/key-value-engine/B-Tree_vs_The_Log-Structured_Merge-Tree.md

```diff
@@ -0,0 +1,101 @@
+
+[B-tree](https://en.wikipedia.org/wiki/B-tree) and [The Log-Structured Merge-tree](https://en.wikipedia.org/wiki/Log-structured_merge-tree) (LSM-tree) are the two most widely used data structures for data-intensive applications to organize and store data. However, each of them has its own advantages and disadvantages. This article aims to use the quantitative approaches to compare these two data structures.
+
+## Metrics
+
+In general, there are three most important metrics to measure the performance of a data structure, which include write amplification, read amplification, and space amplification. This section aims to describe these metrics.
+
+### Write Amplification
+
+`Write amplification` is the ratio of the amount of data written to the storage medium versus the amount of data written to the database. 
+
+For example, if you are writing 10 MB to the database and you observe 30 MB disk write rate, your write amplification is 3.
+
+Flash memory and solid-state drives (SSDs) can be written to only a finite number of times, so that write amplification will decrease the flash lifetime.
+
+There is another write amplification phenomenon associated with the flash memory and SSDs because flash memory must be erased before it can be rewritten.
+
+### Read Amplification
+
+`Read amplification` is the number of disk reads per query. 
+
+For example, if you need to read 5 pages to answer a query, read amplification is 5.
+
+### Space Amplification
+
+`Space amplification` is the ratio of the amount of data on the storage medium versus the amount of data on the database. 
+
+For example, if you put 10MB in the database and this database uses 100MB on the disk, then the space amplification is 10.
+
+## Analysis
+
+The term B-tree may refer to a specific design or a general class of designs. In the narrow sense, a B-tree stores keys in its internal nodes but need not store those keys in the records at the leaves. The [B+ tree](https://en.wikipedia.org/wiki/B%2B_tree#Insertion) is one of the most famous variations of B-tree, behind which the idea is that internal nodes only contain keys, and to which an additional level which contains values is added at the bottom with linked leaves.
+
+LSM-tree performs `compaction` to merge several `sstable`s into one new `sstable` which contains only the live data from the input `sstable`s. Compaction helps LSM-tree to recycle space and reduce read amplification. There are two kinds of `compaction strategy` which is `Size-Tiered compaction strategy` (STCS) and `Level-Based compaction strategy` (LBCS). The idea behind STCS is compacting small `sstable`s into medium `sstable`s when LSM-tree has enough small `sstable`s and compacting medium `sstable`s into large `sstable`s when LSM-tree has enough medium `sstable`s. The idea of LBCS is to organize data into levels and each level contains one sorted run. Once a level accumulates enough data, some of data at this level will be compacted to the higher level.
```

Comment:
I suggest dropping the `SSTable` concepts here because LSM-Tree is a more general storage model.

Reply:
The terminology `sstable` means `Sorted Strings Table` and it doesn't mean any specific data structure.